controllers/topology/reconcile_state.go

```diff
@@ -34,24 +39,88 @@ func (r *ClusterReconciler) reconcileState(ctx context.Context, current, desired
+	if err := r.reconcileControlPlane(ctx, current, desired); err != nil {
+		return err
+	}
 
 	// Reconcile desired state of the InfrastructureCluster object.
 	if err := r.reconcileCluster(ctx, current, desired); err != nil {
 		return err
 	}
 
-	// TODO: reconcile machine deployments
-
-	return nil
+	// Reconcile desired state of the MachineDeployment objects.
+	return r.reconcileMachineDeployments(ctx, current, desired)
 }
 
 // reconcileInfrastructureCluster reconciles the desired state of the InfrastructureCluster object.
 func (r *ClusterReconciler) reconcileInfrastructureCluster(ctx context.Context, current, desired *clusterTopologyState) error {
 	return r.reconcileReferencedObject(ctx, current.infrastructureCluster, desired.infrastructureCluster)
 }
 
+// reconcileControlPlane works to bring the current state of a managed topology in line with the desired state. This involves
+// updating the cluster where needed.
+func (r *ClusterReconciler) reconcileControlPlane(ctx context.Context, current, desired *clusterTopologyState) error {
+	log := ctrl.LoggerFrom(ctx)
+	// Set a default nil return function for the cleanup operation.
+	cleanup := func() error { return nil }
+
+	// If the ControlPlane requires an InfrastructureMachineTemplate
+	// TODO: Check to ensure this is in the clusterclass - make this a common function.
+	// There's a possible issue with using get here - getControlPlaneState uses a class passed to it.
+	// This function also adds a side effect as it's a network call to an outside component - makes testing much more difficult.
+	if ok, err := r.classRequiresInfrastructureMachineTemplate(ctx, desired); err == nil && ok {
+		cpInfraRef, err := getNestedRef(desired.controlPlane.object, "spec", "machineTemplate", "infrastructureRef")
+		if err != nil {
+			return errors.Wrapf(err, "failed to update the %s object,", desired.controlPlane.infrastructureMachineTemplate.GetKind())
+		}
+
+		// Create or update the MachineInfrastructureTemplate fo the control plane.
+		log.Info("updating", desired.controlPlane.infrastructureMachineTemplate.GroupVersionKind().String(), desired.controlPlane.infrastructureMachineTemplate.GetName())
+		cleanup, err = r.reconcileReferencedTemplate(ctx, reconcileReferencedTemplateInput{
+			ref:                  cpInfraRef,
+			current:              current.controlPlane.infrastructureMachineTemplate,
+			desired:              desired.controlPlane.infrastructureMachineTemplate,
+			compatibilityChecker: checkReferencedObjectsAreCompatible,
+			// TODO: Use namer from PR 5070
+			templateNamer: func() string {
+				return names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-infrastructure", desired.controlPlane.infrastructureMachineTemplate.GetName()))
+			},
+		},
+		)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update the %s object", desired.controlPlane.infrastructureMachineTemplate.GetKind())
+		}
+	}
+	// Create or update a ControlPlaneObject for the controlPlaneTopologyState.
```

Comment:
Since the cpInfraRef can be potentailly changed (if changes were detected in the template we are creating a new template) we should make sure to update the machine infraRef in the contorl plane object here with
```
setNestedRef(desired.controlPlane.object, cpInfraRef, "spec", "machineTemplate", "infrastructureRef")
```

Reply:
Context: `getNestedRef` does not get the pointer to the underlying ref in the object. It seems to create a new ObjectReference and set the values accordingly. So, if I am not wrong, updating the values in `cpInfraRef` will not updated the values in the controlPlaneObject. The ref needs to be updated manually.

Reply:
Thanks for catching this! It's a tough one to test for, so I've ended up checking if the resulting reference has a name like those resulting from TemplateRotation. If you see a better way to go about this please let me know.